File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/common/commandbase/ConditionalCommand.java
package org.firstinspires.ftc.teamcode.common.commandbase;

import androidx.annotation.NonNull;

/**
 * Usage:
 * If using a boolean returning function:
 * new ConditionalCommand(() -> {return class.method();});
 */
public class ConditionalCommand extends Command {
    private final LambdaFunction<Boolean> function;

    /**
     * If using a boolean returning function:
     * @param function The function to run
     */
    public ConditionalCommand (LambdaFunction<Boolean> function) {
        this.function = function;
    }

    /**
     * If using a boolean returning function:
     * @param commandName The name of the command
     * @param function The function to run
     */
    public ConditionalCommand(String commandName, LambdaFunction<Boolean> function) {
        super(commandName);
        this.function = function;
    }

    @Override
    public boolean run() {
        return function.run();
    }

    @NonNull
    @Override
    public Command clone() {
        return new ConditionalCommand(getCommandName(), function);
    }
}

File: TeamCode/src/main/kotlin/brickbot/quickstart/commandbase/ConditionalCommand.kt
package brickbot.quickstart.commandbase

import java.util.function.BooleanSupplier

class ConditionalCommand @JvmOverloads constructor(
    commandName: String = "",
    private val booleanFunction: BooleanSupplier
): Command(commandName) {

    override fun run(): Boolean {
        return booleanFunction.asBoolean
    }

    override fun clone(): Command {
        return ConditionalCommand(commandName, booleanFunction)
    }
}
